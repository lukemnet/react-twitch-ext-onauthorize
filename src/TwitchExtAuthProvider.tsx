import React, { ReactElement, useState, useEffect } from 'react';

export interface TwitchAuthResponse {
  channelId: string;
  clientId: string;
  token: string;
  userId: string;
}

export interface TwitchAuthObject extends TwitchAuthResponse {
  authorized: boolean;
}

export interface TwitchAuthProviderProps {
  unauthorized?: ReactElement,
  enable?: boolean,
  children: (auth: TwitchAuthObject) => ReactElement
}

interface TwitchContext extends Window {
  [key: string]: any;
  Twitch: {
    [key:string]: any;
    ext: {
      [key:string]: any;
      onAuthorized: (callback: CallableFunction) => void;
    }
  }
}

const TwitchAuthProvider = (props: TwitchAuthProviderProps) => {
  const enable = (typeof props.enable !== 'undefined') ? props.enable : true;
  const unauthorized = props.unauthorized || <div>Unauthorized</div>;
  const [auth, setAuth] = useState({ authorized: false, channelId: '', token: '', clientId: '', userId: '' });

  useEffect(() => {
    const getTwitchExtensionContext = (callback: CallableFunction) => {
      if (enable) {
        const twitchContext = window as TwitchContext;
        const itsTwitch = !!(twitchContext.Twitch && twitchContext.Twitch.ext);
        if (itsTwitch) {
          twitchContext.Twitch.ext.onAuthorized((twitchAuth: TwitchAuthResponse) => {
            callback({ authorized: true, ...twitchAuth })
          });
        }
      } else {
        callback({
          authorized: true,
          channelId: 'disabled',
          clientId: 'disabled',
          token: 'disabled',
          userId: 'disabled',
        });
      }
    }

    getTwitchExtensionContext((twitchAuthObject: TwitchAuthObject) => {
      setAuth(twitchAuthObject);
    });
  }, [enable]);
  
  return auth.authorized
    ? props.children(auth)
    : unauthorized;
}

export default TwitchAuthProvider;