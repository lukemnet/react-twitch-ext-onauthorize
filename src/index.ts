import TwitchAuthProvider from './TwitchAuthProvider';
export * from './TwitchAuthProvider';
export default TwitchAuthProvider;
