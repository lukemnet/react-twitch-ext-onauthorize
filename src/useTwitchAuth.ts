import { useState, useEffect } from 'react';

export interface TwitchAuthResponse {
  channelId: string;
  clientId: string;
  token: string;
  userId: string;
}

export interface TwitchAuthObject extends TwitchAuthResponse {
  authorized: boolean;
}

interface TwitchContext extends Window {
  Twitch: {
    ext: {
      onAuthorized: (callback: CallableFunction) => void;
    };
  };
}

export const useTwitchAuth = () => {
  const [twitchAuth, setTwitchAuth] = useState({
    authorized: false,
    channelId: '',
    clientId: '',
    token: '',
    userId: '',
  });

  useEffect(() => {
    const twitchContext = (window as Window) as TwitchContext;
    const itsTwitch = twitchContext.Twitch?.ext;
    if (itsTwitch) {
      twitchContext.Twitch.ext.onAuthorized((twitchResponse: TwitchAuthResponse) => {
        setTwitchAuth({ authorized: true, ...twitchResponse });
      });
    }
  },
  [window]);

  return twitchAuth;
};
