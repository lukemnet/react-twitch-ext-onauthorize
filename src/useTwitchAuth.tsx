import { useState, useEffect } from 'react';

export interface TwitchAuthResponse {
  channelId: string;
  clientId: string;
  token: string;
  userId: string;
}

export interface TwitchAuthObject extends TwitchAuthResponse {
  authorized: boolean;
}

interface TwitchContext extends Window {
  [key: string]: unknown;
  Twitch: {
    [key:string]: unknown;
    ext: {
      [key:string]: unknown;
      onAuthorized: (callback: CallableFunction) => void;
    }
  }
}

const useTwitchAuth = () => {
  const [twitchAuth, setTwitchAuth] = useState({
    authorized: false,
    channelId: '',
    clientId: '',
    token: '',
    userId: '',
  });

  useEffect(() => {
    const twitchContext = (window as Window) as TwitchContext;
    const itsTwitch = twitchContext.Twitch?.ext;
    if (itsTwitch) {
      twitchContext.Twitch.ext.onAuthorized((twitchAuth: TwitchAuthResponse) => {
        setTwitchAuth({ authorized: true, ...twitchAuth })
      });
    }
  }, []);

  return twitchAuth;
}

export default useTwitchAuth;