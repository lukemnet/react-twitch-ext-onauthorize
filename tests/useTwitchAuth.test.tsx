import React from "react";
import { render } from "@testing-library/react";
import { useTwitchAuth } from "../src/useTwitchAuth";

const TestElement = () => {
  const twitchAuth = useTwitchAuth();
  return <div>TestElement {JSON.stringify(twitchAuth)}</div>;
};

const mockTwitchExtensionContext = () => {
  Object.defineProperty(window, "Twitch", {
    value: {
      ext: {
        onAuthorized: jest.fn().mockImplementation((callback: Function) => {
          callback({
            authorized: true,
            channelId: "test_channel_id",
            token: "test_token",
            clientId: "test_client_id",
            userId: "test_user_id",
          });
        }),
      },
    },
    writable: true,
  });
};

const resetTwitchExtensionContext = () => {
  (window as any).Twitch = null;
};

describe("useTwitchAuth", () => {
  it("should return Twitch authorization object when authorization passes", () => {
    mockTwitchExtensionContext();
    const { container } = render(<TestElement />);
    expect(container).toMatchSnapshot();
    resetTwitchExtensionContext();
  });

  it("should return object with empty values outside of a Twitch Extension context", () => {
    const { container } = render(<TestElement />);
    expect(container).toMatchSnapshot();
  });
});
