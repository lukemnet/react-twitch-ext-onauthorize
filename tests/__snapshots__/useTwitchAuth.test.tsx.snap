// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTwitchAuth should return Twitch authorization object when authorization passes 1`] = `
<div>
  <div>
    TestElement 
    {"authorized":true,"channelId":"test_channel_id","token":"test_token","clientId":"test_client_id","userId":"test_user_id"}
  </div>
</div>
`;

exports[`useTwitchAuth should return object with empty values outside of a Twitch Extension context 1`] = `
<div>
  <div>
    TestElement 
    {"authorized":false,"channelId":"","clientId":"","token":"","userId":""}
  </div>
</div>
`;
